Owned CRL representation and associated updates.

cpu opened the pull request:
## Description

This branch is the result of [discussion in Discord](https://discord.com/channels/976380008299917365/1121805359452008480) that came about after trying to integrate the existing CRL support added to Webpki with Rustls.

That integration work uncovered a number of things that required a change in approach on the webpki side. For example:

* The `CrlProvider` trait, and working exclusively with CRL representations that borrow from DER didn't mesh well with the Rustls configuration/verifier implementations that have a static lifetime.
* The borrowed CRL representation precluded optimizations that could address needing to do a linear scan of CRL contents to find revoked certificates (https://github.com/rustls/webpki/issues/80).

By dropping the `CrlProvider` trait and providing CRLs directly the lifetimes are much easier to understand. This also better matches how intermediates/trust anchors are provided, and integrates well with Rustls' existing code.

By introducing an owned representation of the CRL types we offer downstream code the option of more performant CRL lookups at the cost of requiring allocation. Most of the owned representation is backed by the borrowed representation, which remains available for users that require webpki to avoid allocation but still want to support revocation checking.

The following are per-commit descriptions of the work involved:

### crl: CertRevocationList -> BorrowedCertRevocationList.

Preparing for introducing a trait, and an owned type.

### verify_cert: drop RevocationOptions/CrlProvider.
In place of the `RevocationOptions` struct and the `CrlProvider` trait, this commit moves to a model where CRLs are provided directly to the verify operation.

The `RevocationOptions` struct was no longer useful after the depth parameter was removed, it simply wrapped a `CrlProvider`.

The `CrlProvider` trait was well intentioned but makes it impossible for a downstream user to operate with owned types, or references that have a lifetime matching the overall lifetime of the validation process.

Passing references to CRLs as part of the verify arguments is much better aligned with how intermediate and trust anchors are provided and will be easier to integrate in Rustls.

The test generator and generated test cases are updated to reflect this new methodology.

### verify_cert: disambiguate CRL signature errors.
This commit introduces CRL specific error variants for errors that can occur during signature validation:

* InvalidCrlSignatureForPublicKey
* UnsupportedCrlSignatureAlgorithm
* UnsupportedCrlSignatureAlgorithmForPublicKey

The CRL validation code is update to return these more specific errors  to avoid confusion about whether a signature error was from validating a certificate in the path to a trust anchor, or from validating a CRL.

### crl: avoid up-front revoked cert iteration.
This commit updates the `BorrowedCertRevocationList` implementation to  avoid iterating and parsing all of the revoked certificate entries when parsing the outer CRL. This allows downstream users more control over when these two parts of the process occur. For example, an owned representation may want to do this iteration up-front to populate a hashmap of serial->revoked cert while other consumers may want to avoid front-loading this work.

To compensate for the fact that we will now have the potential for malformed revoked certificates to be present in the CRL post-parse, the `find_serial` function is updated to return `Result<Option<RevokedCert>>` - this better matches the `into_iter()`
impl and lets the consumer decide how to handle the error case. 

### crl: reorder CRL fields, update visibility.
This commit updates the `BorrowedCertRevocationList` field visibility. In subsequent commits we will expose some of this data via trait methods.

Fields are then sorted by visibility, putting the `pub(crate)` field first. Other fields are made private for now - they aren't used by the webpki or rustls code for verifying client certificates. In the case of the `this_update` and `next_update` fields we allow dead code since these will not be read by anything but may be something we surface in the future.

### crl: add CertRevocationList trait.
In preparation for having both a borrowed representation and an owned representation of CRLs this commit adds a `CertRevocationList` trait that offers the functionality required for webpki to use a CRL to check revocation status of a client certificate.

The `end_entity.rs` and `verify_cert.rs` interfaces for validating a client certificate now accept a `&[&dyn CertRevocationList]` instead of a slice of refs to the borrowed type specifically.

The trait is implemented for the existing `BorrowedCertRevocationList` and the relevant tests are updated. We seal the trait with a private module in `crl.rs` to limit which types can implement it.

### crl: RevokedCert -> BorrowedRevokedCert.
This commit renames the `RevokedCert` type to `BorrowedRevokedCert` in anticipation of also having an owned variant.

### signed_data: add OwnedSignedData.
To support having an owned representation of a CRL we also require an owned representation of `SignedData`. This commit adds that implementation and helper methods for moving between the two.

Since the owned variants and helper methods aren't yet used some temporary `allow(unused)` markers are added to keep the commit building cleanly.

### crl: introduce owned CRL representation.
This commit adds a new `OwnedCertRevocationList` type to accompany the existing `BorrowedCertRevocationList`. To support this we also add a `OwnedRevokedCert` to accompany the existing `BorrowedRevokedCert`. Helper methods to convert from borrowed to owned representation are included.

This representation allows for optimizations like building a `HashMap` from revoked cert serial to owned revoked cert, allowing fast lookup of revoked cert status by serial.

Since the owned representations require allocation they are only made available when the alloc feature has been enabled.

cpu commented on src/crl.rs:
```diff
@@ -37,7 +41,65 @@ pub trait CertRevocationList: private::Sealed {
+///
+/// [^1]: <https://www.rfc-editor.org/rfc/rfc5280#section-5>
+#[cfg(feature = "alloc")]
+pub struct OwnedCertRevocationList {
```
The reason I avoided that was because it bumps into the trait: `CertRevocationList`. Do you have a suggestion for renaming that?

djc replied:
Okay, that seems like a good enough reason! Sorry for the noise.

cpu replied:
No worries :-)

cpu commented on src/crl.rs:
```diff
@@ -37,7 +41,65 @@ pub trait CertRevocationList: private::Sealed {
+#[cfg(feature = "alloc")]
+pub struct OwnedCertRevocationList {
+    /// A monotonically increasing sequence number for a given CRL scope and CRL issuer.
+    pub crl_number: Option<Vec<u8>>,
```
I discussed this in my [comment on the other thread](https://github.com/rustls/webpki/pull/100#discussion_r1243727325).

I thought this information might be useful for consumers that parse a CRL from DER and need to distinguish it from others.

cpu replied:
Visibility adjusted.

djc commented on src/crl.rs:
```diff
@@ -20,28 +20,28 @@ use crate::{der, signed_data, Error, Time};
     pub next_update: Time,
 
+    /// A `SignedData` structure that can be passed to `verify_signed_data`.
+    pub(crate) signed_data: signed_data::SignedData<'a>,
```
Yeah, I think keeping them private is the right default: making them public in a semver-compatible release is trivial, whereas changing anything about a public field is much less so.

cpu replied:
I've made the fields private. I had to add some dead code allows as a result. If we want to avoid those I could rip out the fields + associated parsing entirely. LMK what you prefer.

cpu commented on src/crl.rs:
```diff
@@ -37,7 +41,65 @@ pub trait CertRevocationList: private::Sealed {
+
+    /// A map of the revoked certificates contained in then CRL, keyed by the DER encoding
+    /// of the revoked cert's serial number.
+    pub revoked_certs: HashMap<Vec<u8>, OwnedRevokedCert>,
```
@djc I think there's some nuance here.

Probably we want a newtype over `&[u8]` with a lifetime that could be used both for the `CertRevocationList` trait's `find_serial` arg, and for the return type from `cert.rs`'s `Cert.serial()` fn.

I think we'd need a second owned newtype over `Vec<u8>` for the `OwnedCertRevocationList`'s `revoked_certs` map.

Does that sounds like what you had in mind?

djc replied:
Sounds like the owned newtype might not be worth it.

cpu commented on src/crl.rs:
```diff
@@ -225,17 +219,25 @@ impl<'a> BorrowedCertRevocationList<'a> {
-        self.into_iter()
-            .filter_map(|parse_res| parse_res.ok())
-            .find(|revoked_cert| revoked_cert.serial_number.eq(serial))
+        for revoked_cert_result in self.into_iter() {
```
That was a great tip. Thank you, much nicer :+1:

cpu commented on src/crl.rs:
```diff
@@ -445,6 +473,14 @@ impl TryFrom<u8> for RevocationReason {
 }
 
+mod private {
+    use crate::crl::BorrowedCertRevocationList;
```
I've implemented what I think you were asking for here. Let me know if I missed the mark :-)

cpu commented:
> cpu force-pushed the cpu-crl-support-pt5-owned-type branch from 2269925 to 602028e 

* Integrated all of Djc's feedback (except the serial number new-type, waiting on confirmation there)
* Added a new commit (b2010a91f63ac046e9ffad87aeb0f2a341849678) based on [this discussion on the Rustls side](https://github.com/rustls/rustls/pull/1326#discussion_r1242565148) about making it clearer when signature errors are from CRLs versus certs
* Updated the PR description

djc commented on src/crl.rs:
```diff
@@ -23,25 +23,27 @@ pub struct BorrowedCertRevocationList<'a> {
 
     /// Indicates the issue date of this CRL.
-    pub this_update: Time,
+    #[allow(dead_code)] // May be useful in the future.
```
Unless you have a concrete use case for the near future in mind, I would probably get rid of the unused time fields here. Not like it's a great burden to add them back if/when we need them, right?

cpu replied:
Dropped all the unused/unexposed fields.

djc commented on src/crl.rs:
```diff
@@ -220,27 +243,41 @@ impl<'a> BorrowedCertRevocationList<'a> {
     }
+
+    fn find_serial(&self, serial: &[u8]) -> Result<Option<RevokedCert>, Error> {
+        for revoked_cert_result in self.into_iter() {
```
You probably don't need `into_iter()` here? That's the interface that the language's `for` keyword relies on anyway, so it should call it for you.

cpu replied:
You're right, thanks :+1:

djc commented on src/crl.rs:
```diff
@@ -193,6 +259,41 @@ impl<'a> BorrowedCertRevocationList<'a> {
+        // Parse and collect the CRL's revoked cert entries, ensuring there are no errors. With
+        // the full set in-hand, create a lookup map by serial number for fast revocation checking.
+        let mut revoked_certs = HashMap::new();
+        self.into_iter()
```
I think you should be able to `collect()` into `Result<HashMap<_, _>, _>` directly, by mapping the iterator to a `Result<(K, V), E>` tuple.

cpu replied:
Good call, thanks!
